cusub_perception/detection_tree/src/detection_listener/listener.py

```diff
@@ -0,0 +1,164 @@
+            if self.dobjects[i].class_id == class_id:
+                nums.append(i)
+        return nums
+        
+    def dvector_callback(self, msg): # if we got dobj.num = 1 and len == 1
+        num = msg.dobject_num
+        if num < len(self.dobjects): # We have this dobject
+            self.dobjects[num].add_dvector(msg)
+            self.new_dv_flags[num] = True
+        elif num == len(self.dobjects): # New dobject
+            self.create_new_dobj(num, msg.class_id, msg)
+        else: # Out of order creation of dobject --> we missed a few dvectors
+            rospy.logwarn("We missed a dvector...trigger synchrnoization")
+            return
+
+    def check_new_dv(self, dobj_num):
+        if dobj_num >= len(self.new_dv_flags):
+            return None
+        else:
+            return self.new_dv_flags[dobj_num]
+
+    def clear_new_dv_flag(self, dobj_num):
+        self.new_dv_flags[dobj_num] = False
+
+    def create_new_dobj(self, dobj_num, class_id, dv):
+        self.dobjects.append(Dobject(dobj_num, class_id))
+        self.dobjects[-1].add_dvector(dv)
+        self.new_dv_flags.append(True)
+        
+    # TODO maybe also an option for bearings within a certain distance of current position?
+    def get_avg_bearing(self, dobj_num, num_dv=None, secs=None):
+        """
+        Gets the average bearing over a number of dvectors or time
+
+        Assumes bearings are from the same point (else we'd pose solve)
+
+        Params
+        ------
+        dobj_num : int
+            Which dobject we're talking about
+        num_dv : int (optional)
+            Number of dvectors to average over
+        secs : float (optional)
+            Number of seconds to average over
+        
+        Returns
+        -------
+        [ azimuth, elevation, magnitude ] : float, float, int
+            Average bearing
+        """
+        if num_dv != None and secs != None:
+            rospy.logerr("Incorrect Input 1")
+            return None
+        elif num_dv == None and secs == None:
+            rospy.logerr("Incorrect Input 2")
+            return None
+        elif dobj_num >= len(self.dobjects):
+            rospy.logerr("Incorrect Input 3. Dobject unreceived")
+            return None
+        elif num_dv != None:
+            dvs = self.dobjects[dobj_num].get_most_recent_dvectors(num_dv)
+            return self._get_averages(dvs)
+        else:
+            t_past = rospy.get_rostime() - rospy.Duration.from_sec(secs)
+            dvs = self.dobjects[dobj_num].get_dvectors_since_time(t_past)
+            if dvs == None:
+                return None
+            else:
+                return self._get_averages(dvs)
+
+    def _get_averages(self, dvs):
+        az = np.mean([i.azimuth for i in dvs])
+        el = np.mean([i.elevation for i in dvs])
+        mag = np.mean([i.magnitude for i in dvs])
+        return [az, el, mag]
+
+    def synchronize_dobjects(self): # SHOULDN'T BE NECESSARY IF WE INITIALIZE AT STARTUP
```

Comment:
Would there ever be a situation we would have to maybe clear and restart the detection tree?

Reply:
Right if the perception launch file crashes (like it did at comp) we may have some problems here now that the system is not memoryless